### PR TITLE
Restructuring

### PR DESCRIPTION
**Pull request by Khoroshevskyi**
Everything that was mention in previous PR is here
pepdbagent divided into 5 Modules:
- PEPDatabaseAgent
- BaseConnection
  - PEPDatabaseProject
  - PEPDatabaseAnnotation
  - PEPDatabaseNamespace

Let me know if you have any questions
-----
Few things to add before merging:

- [x] Docs 
- [x] Structured tests
- [x] Add error throwing
- [x] Improve sql performance by reducing LIKE statements and creating indexing

### CODE REVIEW

**Comment by nleroy917** [comment]
So, are we going to close the other PR?

File: pepdbagent/base_connection.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+        :param password: password used to authenticate.
+        :param dsn: libpq connection string using the dsn parameter
+        (e.g. "localhost://username:password@pdp_db:5432")
+        """
+
+        _LOGGER.info(f"Initializing connection to {database}...")
+
+        if dsn is not None:
+            self.pg_connection = psycopg2.connect(dsn)
+            self.db_name = urlparse(dsn).path[1:]
```
Comment (nleroy917) [inline pepdbagent/base_connection.py:40]: It's not immediately clear what this does - maybe move to a helper function with a descriptive name?

File: pepdbagent/base_connection.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        """
+        self.pg_connection.commit()
+
+    def close_connection(self) -> None:
+        """
+        Close connection with database
+        """
+        self.pg_connection.close()
+
+    def __exit__(self):
```
Comment (nleroy917) [inline pepdbagent/base_connection.py:69]: Do we need a `__del__` function that will close the connection when the object gets garbage collected?

File: pepdbagent/const.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-DB_TABLE_NAME = "projects"
+DB_TABLE_NAME = "project"
```
Comment (nleroy917) [inline pepdbagent/const.py:1]: Why was this changed, our of curiosity?

File: pepdbagent/modules/annotation.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+            - provide query and namespace. Return:  list of projects annotations find in specific namespace
+                that have query pattern.
+        :param namespace: Namespace
+        :param name: Project name
+        :param tag: tag
+        :param query: query (search string): Pattern of name, tag or description
+        :param admin: admin name (namespace), or list of namespaces, where user is admin
+        :param limit: return limit
+        :param offset: return offset
+        :return: pydantic model: AnnotationReturnModel(
```
Comment (nleroy917) [inline pepdbagent/modules/annotation.py:67]: Do we need the entire pydantic model inside the doc string? We might be able to remove this, since it's documented using the `->` return syntax.

File: pepdbagent/modules/annotation.py [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+            )
+        """
+        if isinstance(registry_paths, list):
+            anno_results = []
+            for path in registry_paths:
+                try:
+                    namespace, name, tag = registry_path_converter(path)
+                except RegistryPathError as err:
+                    _LOGGER.error(str(err), registry_paths)
+                    continue
```
Comment (nleroy917) [inline pepdbagent/modules/annotation.py:149]: So, this comes up a lot in pephub, and this is a good example. Sometimes when we submit requests to pephub, it appears that they were successful when they were not since `pepdbagent` catches errors and just continues.

One example is project submission or project finding. If you go to pephub and try to submit a project with the same name, `pepdbagent` will catch the error (I believe) and pephub just keeps going without knowing what happened.

So, I think we should throw Exceptions when users make a mistake like supplying an incorrect registry path, if that makes sense. Otherwise it can lead to people thinking something was successful when it was not.

File: pepdbagent/modules/annotation.py [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+                offset=len(registry_paths),
+                result=anno_results,
+            )
+
+        else:
+            try:
+                namespace, name, tag = registry_path_converter(registry_paths)
+            except RegistryPathError as err:
+                _LOGGER.error(str(err), registry_paths)
+                return AnnotationReturnModel()
```
Comment (nleroy917) [inline pepdbagent/modules/annotation.py:166]: Again, here I think just throwing the exception is better. If I supply `nleroy917/basic/basic:latest` I want en error not an empty Annotation object.

**Review by nleroy917 — CHANGES_REQUESTED** [review]
This looks good. I think the API will be really clean and nice to use.

One thing is that there are too many places where we catch errors and return just empty objects. I think "failing gracefully" is really important in software, but for instances like fetching projects, if I ask for a project that doesn't exist, I'd expect an error, not an empty project - if that makes sense.

On the pephub side this is also true, if a user requests a resource that doesn't exist, they get a `404 Not Found` - which is an error code indicating the user messed up. I wouldn't just return an empty list. Maybe @nsheff can weigh in too, but I would err on throwing exceptions more than just catching them.

Otherwise, I am excited to integrate into pephub. It will really improve the development experience.

File: pepdbagent/const.py (hunk 3 above)
Comment (Khoroshevskyi) [reply pepdbagent/const.py:1]: oops, it shouldn't be change! Thanks for this catch!

File: pepdbagent/modules/annotation.py (hunk 4 above)
Comment (Khoroshevskyi) [reply pepdbagent/modules/annotation.py:67]: Yes.. But sometimes it can't be so obvious. What do you think @nsheff?

File: pepdbagent/base_connection.py (hunk 2 above)
Comment (Khoroshevskyi) [reply pepdbagent/base_connection.py:69]: isn't exit doing the same thing? if not  I will add del

File: pepdbagent/base_connection.py (hunk 1 above)
Comment (Khoroshevskyi) [reply pepdbagent/base_connection.py:40]: all class ?

File: pepdbagent/modules/annotation.py (hunk 5 above)
Comment (Khoroshevskyi) [reply pepdbagent/modules/annotation.py:149]: Throwing an error is the easiest solution and I would add it everywhere, but I don't think we should do it.
why? -> Because we don't want to close connection. I know that connection is reopening every time when we call some function, but I would avoid closing connection.
I think solution is: standardize error message ¯\_(ツ)_/¯

Comment (nsheff) [reply pepdbagent/modules/annotation.py:149]: I agree with Nathan. pepdbagent should just be throwing errors. It's at the PEPhub level that they should be caught.,

If pephub catches the error, it doesn't mean the connection will be closed.

File: pepdbagent/modules/annotation.py (hunk 4 above)
Comment (nsheff) [reply pepdbagent/modules/annotation.py:67]: I would probably not include the entire model in the docstring; what if that model changes? the docstring will then be wrong.

The model should be specified in only one place.

Comment (nsheff) [reply pepdbagent/modules/annotation.py:67]: And, to be clear: that one place is where the model is defined. This docstring should just refer to that model.

Comment (Khoroshevskyi) [reply pepdbagent/modules/annotation.py:67]: ok, I agree, I will change it:)

File: pepdbagent/modules/annotation.py (hunk 5 above)
Comment (nsheff) [reply pepdbagent/modules/annotation.py:149]: @Khoroshevskyi I think you shouldn't think of pepdbagent as the final consumer, who must keep running in case of errors. You should think of it as a low-level connection. Throw lots of specific types of Errors that can be caught and then handled by whatever the calling software is.

PEPhub is the ultimate consumer, so all the catching of the specific errors should be handled there.

File: pepdbagent/modules/annotation.py (hunk 6 above)
Comment (nsheff) [reply pepdbagent/modules/annotation.py:166]: I agree

File: pepdbagent/modules/annotation.py [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+        name: str = None,
+        tag: str = None,
+        admin: Union[List[str], str] = None,
+    ) -> Union[AnnotationModel, None]:
+        """
+        Retrieving project annotation dict by specifying project name
+        :param namespace: project registry_path - will return dict of project annotations
+        :param name: project name in database
+        :param tag: tag of the projects
+        :param admin: list of namespaces where user is admin
```
Comment (nsheff) [inline pepdbagent/modules/annotation.py:181]: Is there a proper way to indicate multiple possible types in the docstring?

File: pepdbagent/modules/annotation.py [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+        return None
+
+    def _count_projects(
+        self,
+        namespace: str,
+        search_str: str = "",
+        admin: Union[str, List[str]] = False,
+    ) -> int:
+        """
+        Get total number of found projects. [This function is related to _find_projects]
```
Comment (nsheff) [inline pepdbagent/modules/annotation.py:242]: Instead of "Get total number of found projects.", Just plain "Count projects" is actually more clear...

File: pepdbagent/modules/annotation.py [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+            and_namespace_sql = f"""AND {NAMESPACE_COL} = %s"""
+            namespace = (namespace,)
+        else:
+            and_namespace_sql = ""
+            namespace = tuple()
+
+        count_sql = f"""
+        select count(*)
+            from {DB_TABLE_NAME} 
+                where ({NAME_COL} ILIKE %s or ({PROJ_COL}->>'description') ILIKE %s or {TAG_COL} ILIKE %s)
```
Comment (nsheff) [inline pepdbagent/modules/annotation.py:260]: Instead of having the `search_str` default to '', you should actually use a different sql query here, because `LIKE` is super slow.

Comment (nsheff) [reply pepdbagent/modules/annotation.py:260]: So to clarify, you should only use LIKE if the user is actually requesting a serach. Set the `search_str` default to `None` and module the sql statement you use based on that value.

File: pepdbagent/base_connection.py (hunk 1 above)
Comment (nleroy917) [reply pepdbagent/base_connection.py:40]: I was just saying makings something like:

```python
self._extract_db_name(dsn: str) -> str:
    return urlparse(dsn).path[1:]
```

File: pepdbagent/modules/annotation.py [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+            and_namespace_sql = f"""AND {NAMESPACE_COL} = %s"""
+            namespace = (namespace,)
+        else:
+            and_namespace_sql = ""
+            namespace = tuple()
+
+        count_sql = f"""
+        select {NAMESPACE_COL}, {NAME_COL}, {TAG_COL}, {N_SAMPLES_COL}, ({PROJ_COL}->>'description'), {DIGEST_COL}, {PRIVATE_COL}, {SUBMISSION_DATE_COL}, {LAST_UPDATE_DATE_COL}
+            from {DB_TABLE_NAME} 
+                where ({NAME_COL} ILIKE %s or ({PROJ_COL}->>'description') ILIKE %s or {TAG_COL} ILIKE %s)
```
Comment (nsheff) [inline pepdbagent/modules/annotation.py:307]: Same thing here -- `LIKE` should be used ONLY when a search string is explicitly provided

File: pepdbagent/modules/namespace.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        :return: list of dict with structure {
+                namespace,
+                number_of_projects,
+                number_of_samples,
+            }
+        """
+        search_str = f"%%{search_str}%%"
+        count_sql = f"""
+        select {NAMESPACE_COL}, COUNT({NAME_COL}), SUM({N_SAMPLES_COL})
+            from {DB_TABLE_NAME} where (({NAMESPACE_COL} ILIKE %s and {PRIVATE_COL} is %s)
```
Comment (nsheff) [inline pepdbagent/modules/namespace.py:87]: And here

File: pepdbagent/base_connection.py (hunk 2 above)
Comment (nleroy917) [reply pepdbagent/base_connection.py:69]: Im pretty sure, `__exit__` is called in context managers.

```python
with PEPDatabaseAgent as db:
    db.proj.get(...)
```

So, a `__del__` method is nice for cases outside that.

File: pepdbagent/modules/namespace.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        :param search_str: string of symbols, words, keywords to search in the
+            namespace name.
+        :param admin_nsp: tuple of namespaces where project can be retrieved if they are privet
+        :return: number of found namespaces
+        """
+        search_str = f"%%{search_str}%%"
+        count_sql = f"""
+        select COUNT(DISTINCT ({NAMESPACE_COL}))
+            from {DB_TABLE_NAME} where (({NAMESPACE_COL} ILIKE %s and {PRIVATE_COL} = %s)
+                or ({NAMESPACE_COL} ILIKE %s and {NAMESPACE_COL} in %s ))
```
Comment (nsheff) [inline pepdbagent/modules/namespace.py:122]: Here's another constitutive LIKE use

File: pepdbagent/modules/annotation.py (hunk 7 above)
Comment (nleroy917) [reply pepdbagent/modules/annotation.py:181]: Do we use sphinx for this? I think you can link with `or`. From the [docs](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#info-field-lists)

File: pepdbagent/modules/annotation.py (hunk 9 above)
Comment (nsheff) [reply pepdbagent/modules/annotation.py:260]: If you want, you can use `EXPLAIN` to make sure I'm right about this before going to the effort to make the changes...

File: pepdbagent/base_connection.py (hunk 2 above)
Comment (Khoroshevskyi) [reply pepdbagent/base_connection.py:69]: ok, you are right, I didn't know about it. Thanks for this comment. I will add __del__ to our classes

File: pepdbagent/modules/annotation.py (hunk 5 above)
Comment (Khoroshevskyi) [reply pepdbagent/modules/annotation.py:149]: If pepdbagent isn't final consumer it makes perfect sense. Now I understand how it should work. I will add Exceptions to pepdbagent

File: pepdbagent/modules/annotation.py (hunk 9 above)
Comment (Khoroshevskyi) [reply pepdbagent/modules/annotation.py:260]: Thanks for mentioning it. I didn't know about it, and probably we couldn't even observe performance problems, while testing on small dataset. I agree with all your comments, I will change it

File: pepdbagent/modules/annotation.py (hunk 7 above)
Comment (Khoroshevskyi) [reply pepdbagent/modules/annotation.py:181]: I was using basic restructuredtext docstring style, but without specifying return value 😬

File: docs/README.md [hunk 13]
```diff
@@ -0,0 +1,87 @@
+# pepdbagent
+
+`pepdbagent` is a Python package for uploading, updating, and retrieving PEP metadata from a Postgres database.
```
Comment (nleroy917) [inline docs/README.md:3]: Link to PEP specification?

File: docs/README.md [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+
+Example:
+Set connection to database
+```python
+
+import pepdbagent
+# 1) By providing credentials and connection information:
+agent = pepdbagent.PEPDatabaseAgent(user="postgres", password="docker", )
+# 2) or By providing connection string:
+agent = pepdbagent.PEPDatabaseAgent("postgresql://postgres:docker@localhost:5432/pep-db")
```
Comment (nleroy917) [inline docs/README.md:22]: You probably meant: `postgresql://docker@localhost:5432/pep-db` right?

File: docs/README.md [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+Project is a module that has 3 main purposes:
+- Submitting/Editing projects (+editing metadata)
+- Retrieving projects
+- Deleting projects
+
+Example:
+
+```python
+import peppy
+prj_obj = peppy.Project("prj_path")
```
Comment (nleroy917) [inline docs/README.md:35]: `/path/to/project_config.yaml` is a lot more clear here I think.

File: manual_tests.py [hunk 16]
```diff
@@ -1,61 +1,89 @@
 # file for manual local tests
+import peppy
```
Comment (nleroy917) [inline manual_tests.py:2]: What are `manual_tests.py`? Are those for yourself? If so maybe we can `.gitignore` them.

File: pepdbagent/modules/annotation.py [hunk 17]
```diff
@@ -0,0 +1,337 @@
...
+        :param name: Project name
+        :param tag: tag
+        :param query: query (search string): Pattern of name, tag or description
+        :param admin: admin name (namespace), or list of namespaces, where user is admin
+        :param limit: return limit
+        :param offset: return offset
+        :return: pydantic model: AnnotationReturnModel
+        """
+        if all([namespace, name, tag]):
+            found_annotations = list(
```
Comment (nleroy917) [inline pepdbagent/modules/annotation.py:70]: You are calling this variable `found_annotations` (plural), while calling `_get_single_annotation`. Is this intended?

**Review by nleroy917 — CHANGES_REQUESTED** [review]

File: pepdbagent/modules/annotation.py [hunk 18]
```diff
@@ -0,0 +1,337 @@
...
+                    {PRIVATE_COL},
+                    {PROJ_COL}->>'description',
+                    {N_SAMPLES_COL},
+                    {SUBMISSION_DATE_COL},
+                    {LAST_UPDATE_DATE_COL},
+                    {DIGEST_COL}
+                        from {DB_TABLE_NAME}
+                """
+
+        if tag is None:
```
[the remark this answers is not in the document]
Comment (Khoroshevskyi) [reply pepdbagent/modules/annotation.py:185]: In the very first version of pepdbagent I did it, but somehow we decided to change it. I guess, it's the better way how to do it in Python

File: pepdbagent/modules/annotation.py [hunk 19]
```diff
@@ -0,0 +1,337 @@
...
+    While using this class, user can retrieve all necessary metadata about PEPs
+    """
+
+    def __init__(self, con: BaseConnection):
+        """
+        :param con: Connection to db represented by BaseConnection class object
+        """
+        self.con = con
+
+    def get(
```
[the remark this answers is not in the document]
Comment (Khoroshevskyi) [reply pepdbagent/modules/annotation.py:41]: But I think we should add util function that will check name and tag for special characters and if they make sense

File: docs/README.md (hunk 14 above)
Comment (Khoroshevskyi) [reply docs/README.md:22]: Nope, everything is correct, but I forgot to add `dsn=...`

File: manual_tests.py (hunk 16 above)
Comment (Khoroshevskyi) [reply manual_tests.py:2]: I am running this script to check if everything works correctly, especially SQL queries. Because mock tests are not checking it

File: pepdbagent/exceptions.py [hunk 20]
```diff
@@ -1,12 +1,35 @@
...
+    def __init__(self, msg=""):
+        super().__init__(f"""Project does not exist. {msg}""")
+
+
+class ProjectUniqueNameError(PEPDatabaseAgentError):
+    def __init__(self, msg=""):
+        super().__init__(f"""{msg}""")
+
+
+class ProjectNameError(PEPDatabaseAgentError):
```
[the remark this answers is not in the document]
Comment (Khoroshevskyi) [reply pepdbagent/exceptions.py:33]: ProjectNameError -> this error is raised when name was not provided

File: pepdbagent/modules/annotation.py (hunk 17 above)
Comment (Khoroshevskyi) [reply pepdbagent/modules/annotation.py:70]: it's problem with naming...
As project has few annotations, and we call it annotation🤔
It's good point, I will change it to singular

File: pepdbagent/modules/annotation.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Khoroshevskyi) [reply pepdbagent/modules/annotation.py:185]: I think it make no difference. If you want I can change it. The problem would be if argument default value is mutable variable, but we don't have it

File: pepdbagent/exceptions.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (nleroy917) [reply pepdbagent/exceptions.py:33]: I think that should just be `ValueError` then, right? I'm not really sure. I'm more agnostic about all this.

[the remark this answers is not in the document]
Comment (Khoroshevskyi) [reply pepdbagent/exceptions.py:33]: Probably you are right. I will change it. It will add clarity to the code

File: manual_tests.py (hunk 16 above)
Comment (nsheff) [reply manual_tests.py:2]: I like including manual tests in there, it's nice for me. So I'd not .gitignore them.

**Comment by nsheff** [comment]
Hey @Khoroshevskyi  this API is really nice, I really like it. This is way more clear now!

I went through the tutorial and edited it a bit. There are few things that still confused me, and I wasn't sure how to edit it. Can you rephrase these points:

- [ ] "- Submitting/editing projects (+editing metadata)"  (what is "+ editing metadata"? isn't that what annotation does? Please explain
- [ ] can you provide a bit more insight/example onto how `agetn.project.edit` works? is this how you update a project or just its annotation? can you provide some more examples? Why isnt't there `agent.annotation.edit` ?
- [ ]  "retrieves information about namespaces and provides search functionality." -> can you clarify how this differs from the search functionality described earlier? just be a bit more specific about the differences here.
- [ ]  This is not clear to me: "# Additionally, to get annotations from private projects you should provide # admin or list of admins."
- [ ] This is also not clear to me: "# The same way as in annotation you should provide admin list, to have # information about private projects added to information about namespaces"

**Comment by nsheff** [comment]
Here are some ideas for word changes, what do you think?

- change `agent.project.edit` to `agent.project.update`.
- change `agent.project.submit` to `agent.project.create`.

**Comment by nleroy917** [comment]
+1 for word changes @nsheff suggested.

File: pepdbagent/modules/annotation.py [hunk 21]
```diff
@@ -0,0 +1,327 @@
...
+        :param name: Project name
+        :param tag: tag
+        :param query: query (search string): Pattern of name, tag or description
+        :param admin: admin name (namespace), or list of namespaces, where user is admin
+        :param limit: return limit
+        :param offset: return offset
+        :return: pydantic model: AnnotationReturnModel
+        """
+        if all([namespace, name, tag]):
+            found_annotation = list(
```
Comment (nleroy917) [inline pepdbagent/modules/annotation.py:70]: Hey @Khoroshevskyi, I've been testing this with `pephub` and this line is a bug. Calling `list()` on a model defined through `pydantic` does not create a list with one instance of the model in it, it actually creates a list of `tuples` from the key-value pairs in the model.

Everything inherits from the `BaseModel` class - check out the `__iter__` [method](https://github.com/pydantic/pydantic/blob/ef8f9fe51b9d0e38693f70482c29ba124f810d5a/pydantic/main.py#L488) of `BaseModel`. `list()` takes an iterable. I'd suggest just making this line:

```python
            found_annotation = [self._get_single_annotation(
                    namespace=namespace,
                    name=name,
                    tag=tag,
                    admin=admin,
                )
            )]
```

Comment (Khoroshevskyi) [reply pepdbagent/modules/annotation.py:70]: do you mean changing `list()` to `[]`?

Comment (nleroy917) [reply pepdbagent/modules/annotation.py:70]: Yeah. `list()` [takes an iterable](https://docs.python.org/3/library/functions.html#func-list) and generates a list from that.

Comment (Khoroshevskyi) [reply pepdbagent/modules/annotation.py:70]: I see, I just got the same error. Fixing ...

File: docs/README.md (hunk 14 above)
Comment (nleroy917) [reply docs/README.md]: i see now. The username being `postgres` threw me off...

File: pepdbagent/modules/annotation.py (hunk 21 above)
Comment (nsheff) [reply pepdbagent/modules/annotation.py:70]: yes, you are right, you have to use `[]`

**Comment by nsheff** [comment]
"# submit a project" -> "# create a project"

**Comment by nsheff** [comment]
Why is there only a `_by_rp` function for `annotation.get_by_rp`? Do you need functions `create_by_rp` and `update_by_rp`, et. and put them on project, annotation, namespace modules?

**Comment by nsheff** [comment]
Instead of "NamespaceReturnModel" it doesn't make sense to put the word "Return" in this object, because it's not always going to be "Returned" it could be used for lots of other things.

What is this really? Can you name it appropriately?

NamespaceList? Similar for Annotation? `AnnotationList` object maybe?

**Comment by Khoroshevskyi** [comment]
> Why is there only a `_by_rp` function for `annotation.get_by_rp`? 

I don't really understand this question

> Do you need functions `create_by_rp` and `update_by_rp`, et. and put them on project, annotation, namespace modules?

If we need I will add it, but I am not sure if we are using it. https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

**Comment by Khoroshevskyi** [comment]
> Instead of "NamespaceReturnModel" it doesn't make sense to put the word "Return" in this object, because it's not always going to be "Returned" it could be used for lots of other things.
> 
> What is this really? Can you name it appropriately?
> 
> NamespaceList? Similar for Annotation? `AnnotationList` object maybe?

Yes, probably we should use different name for it. Your suggestions?? 
(p.s. In my experience all model naming should end on `Model` so in will be obvious that it's model)

**Comment by nsheff** [comment]
The pydantic docs don't suggest ending every model with `Model`: https://docs.pydantic.dev/usage/models/

I would suggest `NamespaceList` and `AnnotationList` ...

**Comment by nsheff** [comment]
Is this model an Annotation or Annotations ? Can you sync the model name and docstring one way or the other?

https://github.com/pepkit/pepdbagent/blob/9621f05e61e14b11af61e7800ca6c28a2f4288ae/pepdbagent/models.py#L9-L11

**Comment by nsheff** [comment]
Can you justify why you're prefixing double underscores here? I'm not familiar with that convention, can you explain why you're not just using the single underscore approach that I'm familiar with?

https://github.com/pepkit/pepdbagent/blob/9621f05e61e14b11af61e7800ca6c28a2f4288ae/pepdbagent/pepdbagent.py#L37-L43

**Comment by nsheff** [comment]
What's the advantage of attaching an object at `_project` (or `__project`) and then defining a @property called `project` separately?

How is this superior to just sticking the object on `.project` in the first place?

**Comment by Khoroshevskyi** [comment]
> 

1) two underscore means private object, it means that you can't change it outside of the object. (You can but in different way)
2) if I am using property, no one can change `__project` or `__namespace` variable outside pepdbagent

File: pepdbagent/modules/annotation.py [hunk 22]
```diff
@@ -0,0 +1,327 @@
...
+    ) -> AnnotationList:
+        """
+        Get project annotations by providing registry_path or list of registry paths.
+        :param registry_paths: registry path string or list of registry paths
+        :param admin: list of namespaces where user is admin
+                :return: pydantic model: AnnotationReturnModel(
+            limit:
+            offset:
+            count:
+            result: List [AnnotationModel])
```
Comment (nleroy917) [inline pepdbagent/modules/annotation.py:112]: While integrating with pephub, sometimes I will see `result` and sometimes I see `results` (plural). I was wondering if we could standardize that

Comment (Khoroshevskyi) [reply pepdbagent/modules/annotation.py:112]: I did it already 😀

Comment (nleroy917) [reply pepdbagent/modules/annotation.py:112]: Oh good. thats probably why I couldnt find a concrete example lol

**Comment by nleroy917** [comment]
> The pydantic docs don't suggest ending every model with `Model`.

I was going to ask also about the naming of the models. What's the difference between `____ResultModel` and `___ReturnModel`? I was getting them mixed up a bit.

**Comment by Khoroshevskyi** [comment]
> > The pydantic docs don't suggest ending every model with `Model`.
> 
> I was going to ask also about the naming of the models. For example, what's the difference between `NamespaceResultModel` and `NamespaceReturnModel`? I was getting them mixed up a bit.

Yeah, again naming... I changed it already. so now you will have:
NamespaceList: it returns model that has list of NamespaceRetrunModel 
NamespaceResultModel: is ONE result

AnnotationList: it returns model that has list of Annotation
Annotation: is ONE annotation

**Comment by nsheff** [comment]
Can `NamespaceResultModel` just be `Namespace` then, for consistency with `Annotation` ?

**Review by nsheff — APPROVED** [review]
Ok I think this is good enough for now. Nice work! Let's deploy.

**Review by nleroy917 — APPROVED** [review]
